Add apply_polygon #287

m-mohr opened the pull request:
A proposal to discuss for issue #287.

(Not needed in deliverable)

m-mohr commented on proposals/chunk_polygon.json:
```diff
@@ -0,0 +1,80 @@
+            "name": "chunks",
+            "description": "A GeoJSON object containing at least one polygon. The provided feature types can be one of the following:\n\n* A `Polygon` or `MultiPolygon` geometry,\n* a `Feature` with a `Polygon` or `MultiPolygon` geometry,\n* a `FeatureCollection` containing at least one `Feature` with `Polygon` or `MultiPolygon` geometries, or\n* a `GeometryCollection` containing `Polygon` or `Mu ...
+            "schema": {
+                "type": "object",
```
We may add vector-cube as a type here.

m-mohr replied:
I'll do this separately in one go for all processes.

soxofaan commented on proposals/chunk_polygon.json:
```diff
@@ -0,0 +1,80 @@
+        },
+        {
+            "name": "process",
+            "description": "A process that accepts and returns a single data cube and is applied on each individual chunk. The process may consist of multiple sub-processes.",
```
> The process may consist of multiple sub-processes.

Isn't this a bit redundant? It's basically the definition of a process.

soxofaan replied:
I think there should also be a note that the process should preserve the original dimensions (so no adding/dropping of dimensions, no type changes, ...). 

I'm not sure it should be allowed or forbidden to also change labels, cardinality, resolution,

m-mohr replied:
1) Yes for us "pros" that is clear, but we have that in all(?) callback definitions afaik. The reason is that at some point we had people asking whether they could only provide a single process (e.g. "mean") here or more.

2) Good point. I'm not actually sure what the intention here was originally, so forwarding this question to @JeroenVerstraelen and @jdries ...

JeroenVerstraelen replied:
@m-mohr Our implementation assumes that no dimensions can be added/dropped and that the (#rows, #cols) of the input is identical to that of the output. But you can add/drop individual dates or bands.

As it was originally requested in https://github.com/Open-EO/openeo-processes/issues/287, the user needed to be able to view a 4D cube per polygon and create a correlation matrix based on some computation for each date.

I believe that adding/dropping dimensions falls outside of the scope of this process but I think that the preservation of the spatial dimensions can be discussed. Taking into account this documentation for apply https://openeo.org/documentation/1.0/datacubes.html#apply the fact that we call the function 'apply_polygon' could mean:
1. Identical to apply_neighbourhood, apply_kernel: apply a function over every pixel in the datacube using a polygon as a kernel (wrong)
2. Identical to apply_dimension: apply a function over every pixel using the polygon as the input (right)

(In our case we provide a list of polygons so apply_polygons might be an even better name.) So we can use the exact same description as we use for the _apply_ process, it is just that we divided it into an apply for every polygon instead of the whole datacube. The return value of the apply process is:
_A data cube with the newly computed values and the same dimensions. The dimension properties (name, type, labels, reference system and resolution) remain unchanged._

Given that description I think our VITO implementation already permits too much by allowing to add/drop dates or bands.

m-mohr replied:
So is it fair to update the description of the returned sub data cube to

> The updated sub data cube with the newly computed values and the same dimensions. The dimension properties (name, type, reference system and resolution) must remain unchanged. The labels can change, but the number of labels must remain unchanged.

m-mohr commented:
Sad to see that there was no communication in the PR to further discuss and add this to this PR during the implementation phase. Now, I can't guarantee that the definition here will be exactly what you expose and we may diverge.

JeroenVerstraelen commented:
The mask_value parameter was requested by the internal VITO member who asked for this process. 

The issue was that a UDF receives data for a rectangular area, but geometries can have any shape. So any pixels not in the geometry had to be masked out with a value provided by the user.

m-mohr commented:
This proprietary parameter will likely not become part of the spec though as this usecase is already covered by mask_polygon. I'd suggest using that instead of a proprietary addition.

m-mohr commented:
I've just added a definition of mask_value to chunk_polygon. Please review.

soxofaan commented:
It's probably a bit late to raise this, 
but while working on documentation it appeared to me that the name "chunk_polygon" does not align well with comparable processes like `apply`, `apply_dimension`, `apply_neighborhood`, `reduce_dimension`. ...  which contain a verb like "apply" or "reduce" that make it self-documenting.
I think chunk_polygon belongs to the "apply" family, but that is not obvious: usage-wise the process does not really "chunk" your data (in the sense that it would return an array of chunks or something like that);  it "applies" a child process to chunks (internally), but still returns a single data cube.
Is there still opportunity to rename to, e.g. "apply_chunks" , "apply_chunked" or "apply_polygon"?

m-mohr commented:
@soxofaan I think you need to raise this internally, because I think only VITO itself is implementing it.

soxofaan commented:
indeed, I should have cc'ed @jdries .  
Still, as it's a naming/allignment thing, it doesn't hurt to discuss this publicly too.

m-mohr commented:
Sure, but what I meant is whether it's fine for you too change at all. It VITO says you don't want to change for backward compatibility, it doesn't make a lot of sense to discuss it here, too.

jdries commented:
The name can still change, we know most of the cases and persons that use this, and can always keep an alias for backwards compatibility.
Note that the process is quite popular, it's used by:
- electrical conductivity service
- water reservoir mappoing
- geo informed

m-mohr commented:
Okay, sounds good.

Any preferences? Some thoughts:
- This process is either a variant of apply or mask.
- `apply_chunked` and similarly generic names don't reflect the spatial nature. How would we call a chunking process over e.g. the temporal dimension?
- `apply_polygon` would be a short variant that hopefully implies chunking
- `apply_chunks_polygon` could be a bit more clear, but it a mouthful...

soxofaan commented:
I also think  `apply_polygon` is the best candidate. It aligns very well with `apply_neighborhood` in terms of the chunking aspect

m-mohr commented:
Renamed. Would be really nice to finally merge this so requesting reviews again: @jdries @soxofaan @JeroenVerstraelen

jdries commented:
Looks good, is there a need to say something about non-spatial dimensions? Do we need to mention explicitly that chunking happens along the x,y dimensions, and all others are passed on fully into the 'chunk'?

jdries commented:
@JeroenVerstraelen  can you add your review?

m-mohr commented:
@jdries That's probably a good clarificaton to add.

Another question that is still open is https://github.com/Open-EO/openeo-processes/pull/298#discussion_r822546408 @JeroenVerstraelen

LukeWeidenwalker commented on proposals/apply_polygon.json:
```diff
@@ -0,0 +1,100 @@
+{
+    "id": "apply_polygon",
+    "summary": "Apply a process to spatial chunks of a data cube",
+    "description": "The given data cube is chunked by the given polygons and applies the given process to each individual chunk.",
```
This is a nitpick, but as a dask user the word "chunk" is overloaded for me. In dask, it would be possible that two of these polygons live within the same chunk. Also dask chunks are rectangular, which geometries are not necessarily. How about something like "The provided process is applied only to the subregions of the data cube described by the provided polygons."?

soxofaan replied:
Indeed, with the rename to `apply_polygon` we might want to steer away from using "chunk" all together

m-mohr replied:
Hmm... happy to optimize the wording, but "Apply a process to each given polygon in the datacube." sounds for me like you apply something on a vector data cube. I'll think about an alternative...

JeroenVerstraelen replied:
Maybe it was also nitpicky from me. Given that we want to emphasize that the polygons are cut out of the input data cube first and then a process is applied to each polygon the longer sentence is probably better.

LukeWeidenwalker reviewed (commented):
One nitpick and one general comment: 
Is there anything this process should say about what happens with overlapping geometries, or is handling of this entirely the business of backends? I guess since there's no focal aspect to this process, the results will be the same regardless of whether geometries overlap, and it's thus only a potential performance optimisation?

m-mohr commented on proposals/apply_polygon.json:
```diff
@@ -0,0 +1,100 @@
+            }
+        },
+        {
+            "name": "chunks",
```
Good idea, will rename.

m-mohr commented on proposals/apply_polygon.json:
```diff
@@ -0,0 +1,100 @@
+        }
+    ],
+    "returns": {
+        "description": "A data cube with the newly computed values and the same dimensions. The dimension properties (name, type, labels, reference system and resolution) remain unchanged.",
```
Does the general process description that I just updated help to clarify this?

m-mohr commented:
I've updated the descriptions of the process according to the comments and also added vector data cube support. Please let me know what you think, I hope we can approve this soon.

JeroenVerstraelen reviewed (approved):
Looks good to me!

m-mohr commented:
@LukeWeidenwalker @clausmichele @dthiex I'd appreciate your final reviews on this PR.

m-mohr commented:
I'll merge this EOB on Tuesday if no other comments come in.

LukeWeidenwalker commented:
One question still open from me: What's the expected behaviour for pixels that intersect multiple of the provided geometries?

m-mohr commented:
@LukeWeidenwalker Hmm... good question. Open to suggestions. As it is hard to actually get reproducible results if you run over an area multiple times, should we throw an error if overlapping geometries are provided?

LukeWeidenwalker commented:
> @LukeWeidenwalker Hmm... good question. Open to suggestions. As it is hard to actually get reproducible results if you run over an area multiple times, should we throw an error if overlapping geometries are provided?

I think throwing an error is the best option here, doing this multiple times over an area just seems ill-defined to me!

m-mohr commented:
Okay, I've added an exception.

jdries commented:
> Okay, I've added an exception.

If polygons overlap, I believe the function is still invoked once per polygon and pixels are masked according to the polygon that is being handled. Not sure if this is ill defined and requires an exception? @JeroenVerstraelen

m-mohr commented:
I assumed @JeroenVerstraelen thumbs up to Lukas' comment meant agreement.

m-mohr commented:
Great, thanks. So I'll merge this now. If there's further room for improvement, please open a new issue and/or PR.
